_posts/2022-08-17-using-the-kani-rust-verifier-on-tokio-bytes.md

```diff
@@ -0,0 +1,556 @@
+
+    ptr_in_bounds && cap_in_bounds
+}
+```
+
+Phew, that was a bit of work!
+But now we have a property (a representation invariant) that should *always hold* for any instance of `BytesMut`.
+
+### A proof about `with_capacity()`
+
+Let's try and use our representation invariant.
+How about proving that the `BytesMut` constructor `with_capacity()` always returns a well-formed instance?
+In [earlier posts](https://model-checking.github.io/kani-verifier-blog/2022/05/04/announcing-the-kani-rust-verifier-project.html#enter-kani) we introduced Kani [`any<T>()`](https://model-checking.github.io/kani/tutorial-nondeterministic-variables.html).
+This is a feature that informally generates "any `T` value".
+This is not the same as a randomly chosen concrete value (as in fuzzing) but rather a *symbolic* value that represents any possible value of the appropriate type.
+The key idea is that we can use `any` in a Kani harness to verify the behavior of our code with respect to all possible values of an input (rather than having to exhaustively enumerate them).
+For example, we can write a harness that uses `any<usize>()` to call `with_capacity` and subsequently assert that the returned `BytesMut` instance is well-formed.
+
+```rust
+#[cfg(kani)]
+mod verification {
+    use super::*;
+
+    #[kani::proof]
+    fn with_capacity_returns_well_formed_bytes_mut() {
+        let cap = kani::any();
+        kani::assume(0 < cap);                       //< (1)
+        kani::assume(    cap < MAX_KANI_ALLOCATION); //< (2)
+        let a = BytesMut::with_capacity(cap);
+        assert!(a.kind() == KIND_VEC);
+        assert!(a.is_valid());
+    }
+}
+```
+
+There are two assumptions that we will return to shortly, but first we note Kani has no problem proving this harness:
+
+```bash
+$ cargo kani --harness with_capacity_returns_well_formed_bytes_mut --output-format terse
+#--snip--
+VERIFICATION RESULT: 
+ ** 0 of 410 failed
+
+VERIFICATION:- SUCCESSFUL
+```
+
+The two assumptions are needed for different reasons.
+The first assumption `(1)` avoids constructing a vector with capacity `0`.
+This doesn't sound tricky except that the `Vec` implementation uses a dangling pointer to represent vectors with capacity `0`.
+And we have an [open question](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Behavior.20of.20dangling.20pointer.20arithmetic.20operations) with the unsafe-wg about what operations are legal on such a dangling pointer.
+For now, using this assumption avoids this case.
+
+The second assumption`(2)` is necessary because Kani's abstract machine uses a fixed (but parameterizable) number of bits in a pointer for its analysis (to track allocated objects).
+By default, Kani's maximum allocation is `2^47` bytes (whereas the documentation for `vec` states a maximum allocation of `isize::MAX == 2^63 - 1 bytes`).
+Currently Kani does not error if an allocation exceeds this bound so this assumption avoids this case.
```

Comment:
"does not error"?

Reply:
Yeah, it just wraps around. Can you review this part to make sure the numbers are alright?